apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/so11y/AgentSo11y.java

```diff
@@ -0,0 +1,146 @@
+                        .counter("created_tracing_context_counter")
+                        .tag("created_by", "propagated")
+                        .build();
+                }
+                PROPAGATED_CONTEXT_COUNTER.increment(1);
+            }
+        } else {
+            if (ignoredTracingContext) {
+                if (SAMPLER_IGNORE_CONTEXT_COUNTER == null) {
+                    SAMPLER_IGNORE_CONTEXT_COUNTER = MeterFactory
+                        .counter("created_ignored_context_counter")
+                        .tag("created_by", "sampler")
+                        .build();
+                }
+                SAMPLER_IGNORE_CONTEXT_COUNTER.increment(1);
+            } else {
+                if (SAMPLER_CONTEXT_COUNTER == null) {
+                    SAMPLER_CONTEXT_COUNTER = MeterFactory
+                        .counter("created_tracing_context_counter")
+                        .tag("created_by", "sampler")
+                        .build();
+                }
+                SAMPLER_CONTEXT_COUNTER.increment(1);
+            }
+        }
+    }
+
+    public static void measureTracingContextCompletion(boolean ignoredTracingContext) {
+        if (ignoredTracingContext) {
+            if (FINISH_IGNORE_CONTEXT_COUNTER == null) {
+                FINISH_IGNORE_CONTEXT_COUNTER = MeterFactory.counter("finished_ignored_context_counter").build();
+            }
+            FINISH_IGNORE_CONTEXT_COUNTER.increment(1);
+        } else {
+            if (FINISH_CONTEXT_COUNTER == null) {
+                FINISH_CONTEXT_COUNTER = MeterFactory.counter("finished_tracing_context_counter").build();
+            }
+            FINISH_CONTEXT_COUNTER.increment(1);
+        }
+    }
+
+    public static void measureLeakedTracingContext(boolean ignoredTracingContext) {
+        if (ignoredTracingContext) {
+            if (LEAKED_IGNORE_CONTEXT_COUNTER == null) {
+                LEAKED_IGNORE_CONTEXT_COUNTER = MeterFactory
+                    .counter("possible_leaked_context_counter")
+                    .tag("source", "ignore")
+                    .build();
+            }
+            LEAKED_IGNORE_CONTEXT_COUNTER.increment(1);
+        } else {
+            if (LEAKED_CONTEXT_COUNTER == null) {
+                LEAKED_CONTEXT_COUNTER = MeterFactory
+                    .counter("possible_leaked_context_counter")
+                    .tag("source", "tracing")
+                    .build();
+            }
+            LEAKED_CONTEXT_COUNTER.increment(1);
+        }
+    }
+
+    public static void durationOfInterceptor(double timeCostInNanos) {
+        if (INTERCEPTOR_TIME_COST == null) {
+            INTERCEPTOR_TIME_COST = MeterFactory
+                .histogram("tracing_context_performance")
+                .steps(Arrays.asList(0.01d, 0.1d, 0.5d, 1d, 3d, 5d, 10d, 50d, 100d, 200d, 500d, 1000d))
+                .build();
+        }
+        INTERCEPTOR_TIME_COST.addValue(timeCostInNanos / 1000000);
```

Comment:
Metrics should be raw data, right? I don't think a method of interceptor would cost over 1 ms?

Reply:
Do you have concern of long overflow?

Reply:
How does the value look like in your testing?

Reply:
It is a group of time cost(nanos) in `agent-so11y-scenario`, we should update histogram steps too, right?
```
4453.0 
1.5740867E7 
1875.0 
3276344.0 
1923509.0 
2316475.0 
932202.0 
37576.0 
1706292.0 
1846.0 
303985.0 
126229.0 
3075.0 
203682.0 
1919.0 
29810.0 
2636649.0 
```

Reply:
Yes, I think they should be aligned. Consider 4k ns, I think to keep ns as unit makes sense. Agree?

Reply:
Yes, ns as unit make sense.

Reply:
Let's keep the first buckets of histogram as 1000ns.

Reply:
> Let's keep the first buckets of histogram as 1000ns.

Is this step ok?
```
private static final List<Double> TIME_COST_HISTOGRAM_STEPS = Arrays.asList(
    1000d, 10000d, 50000d, 100000d, 300000d, 500000d,
    1000000d, 5000000d, 10000000d, 20000000d, 50000000d, 100000000d
);
```

Reply:
The max is 100ms per intercepting. I think it is a very bad performance.

Let's do this anyway. I am not sure how bad an interceptor would be.